snuba/writer.py

```diff
@@ -39,10 +40,32 @@ def __row_to_column_list(self, columns, row):
 
     def write(self, rows: Iterable[WriterTableRow]):
         columns = self.__schema.get_columns()
-        self.__connection.execute_robust("INSERT INTO %(table)s (%(colnames)s) VALUES" % {
-            'colnames': ", ".join(col.escaped for col in columns),
-            'table': self.__schema.get_table_name(),
-        }, [self.__row_to_column_list(columns, row) for row in rows], types_check=False)
+        self.__connection.execute_robust(
+            "INSERT INTO %(table)s (%(colnames)s) VALUES"
+            % {
+                "colnames": ", ".join(col.escaped for col in columns),
+                "table": self.__schema.get_table_name(),
+            },
+            [self.__row_to_column_list(columns, row) for row in rows],
+            types_check=False,
+        )
+
+
+class ClickHouseError(Exception):
+    def __init__(self, code: int, message: str):
+        self.code = code
+        self.message = message
+
+    def __str__(self) -> str:
+        return f"[{self.code}] {self.message}"
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {self}>"
+
+
+CLICKHOUSE_ERROR_RE = re.compile(
```

Comment:
How confident are we all error have this format ? Is it in the doc or did we find this out empirically ?

Reply:
It's not documented anywhere I could find, so yeah, this is empirical. I'm not exactly sure where it comes from in the C source either.

It's a consistent format across several of the error messages that I saw, though a quick search of Sentry makes it look like we should also use `NetException` in addition to `Exception`. I probably should just capture group the exception type.

I'm hoping that we'll be able to switch to the JSON error format in the near-ish future.

Reply:
Added the type as part of the capture group with ddf6ff44f26021adbbc8a25007de85a39e1a6d80.

Reply:
Maybe you could add the raw payload of the error to the exception if it does not respects the regex just to stay on the safe side.